File: PullNewController/PullNewController/TabBarController.m
//
//  TabBarController.m
//  PullNewController
//
//  Created by liyufeng on 2019/5/26.
//  Copyright © 2019 liyufeng. All rights reserved.
//

#import "TabBarController.h"
#import "FirstViewController.h"
#import "SecondViewController.h"
#import "MainPullViewController.h"
#import "ListViewController.h"

@interface TabBarController ()<PullContainerViewControllerDelegate>

@property (nonatomic, assign)CGRect tabBarOriginFrame;
@property (nonatomic, assign)BOOL willTabBarHidden;

@end

@implementation TabBarController

- (void)viewDidLoad {
    [super viewDidLoad];
    UIViewController * top = [[ListViewController alloc]initWithNibName:nil bundle:nil];
    UIViewController * main = [[FirstViewController alloc]initWithNibName:nil bundle:nil];
    PullContainerViewController * pullController = [[MainPullViewController alloc]initWithTopView:top andMainView:main];
    [pullController.delegates addPointer:(void * _Nullable)(self)];
    
    CGFloat statusHeight = CGRectGetHeight([UIApplication sharedApplication].statusBarFrame);
    pullController.middle = ([UIScreen mainScreen].bounds.size.height-statusHeight-44)/3;
    pullController.max = ([UIScreen mainScreen].bounds.size.height-statusHeight-44);
    pullController.tabBarItem = [[UITabBarItem alloc]initWithTitle:@"pull" image:nil tag:0];
    NSMutableArray * arr = [NSMutableArray arrayWithArray:self.viewControllers];
    [arr addObject:pullController];
    [self setViewControllers:[NSArray arrayWithArray:arr] animated:YES];
}

- (void)viewWillAppear:(BOOL)animated{
    [super viewWillAppear:animated];
}

- (void)viewDidAppear:(BOOL)animated{
    [super viewDidAppear:animated];
    self.tabBarOriginFrame = self.tabBar.frame;
    self.willTabBarHidden = self.tabBar.hidden;
}

- (void)setTabbarHidden:(BOOL)hidden animation:(BOOL)animation{
    if (self.tabBar.hidden == hidden) {
        return;
    }
    if (self.willTabBarHidden == hidden) {
        return;
    }
    self.willTabBarHidden = hidden;
    __weak typeof(self)weakSelf = self;
    void(^handle)(void) = ^{
        if (hidden) {
            CGRect frame = self.tabBar.frame;
            weakSelf.tabBar.frame = CGRectMake(frame.origin.x, self.tabBarOriginFrame.origin.y+frame.size.height, frame.size.width, frame.size.height);
            weakSelf.tabBar.alpha = 0;
        }else{
            CGRect frame = self.tabBar.frame;
            weakSelf.tabBar.frame = CGRectMake(frame.origin.x, self.tabBarOriginFrame.origin.y, frame.size.width, frame.size.height);
            weakSelf.tabBar.alpha = 1;
        }
    };
    if (animation) {
        [UIView animateWithDuration:0.1 animations:^{
            handle();
        } completion:^(BOOL finished) {
            weakSelf.tabBar.hidden = hidden;
        }];
    }else{
        handle();
        weakSelf.tabBar.hidden = hidden;
    }
}

- (void)pullContainerViewController:(PullContainerViewController *)controller progress:(CGFloat)progress{
    if (controller.willChangeToStatus == PullContainerViewControllerStatusShowTop) {
        if (!self.tabBar.hidden) {
            [self setTabbarHidden:YES animation:YES];
        }
    }else{
        if (self.tabBar.hidden) {
            [self setTabbarHidden:NO animation:YES];
        }
    }
}

@end

File: PullViewController-Swift/PullViewController-Swift/controller/TabBarController.swift
//
//  TabBarController.swift
//  PullViewController-Swift
//
//  Created by liyufeng on 2019/5/27.
//  Copyright © 2019 liyufeng. All rights reserved.
//

import UIKit

@objc class TabBarController: UITabBarController {
    var tabBarOriginFrame : CGRect = CGRect.zero
    var willTabBarHidden : Bool = false
    
    override func viewDidLoad() {
        super.viewDidLoad()
        let main = FirstViewController()
        let top = SecondViewController()
        let pullController = PullContainerViewController(topVc: top, mainVc: main)
        pullController.delegates.addPointer(Unmanaged.passRetained(self).toOpaque())
        pullController.delegates.addPointer(Unmanaged.passRetained(top).toOpaque())
        pullController.delegates.addPointer(Unmanaged.passRetained(main).toOpaque())
        pullController.middle = (UIScreen.main.bounds.size.height-83)/3
        pullController.max = (UIScreen.main.bounds.size.height-83)
        pullController.tabBarItem = UITabBarItem(title: "pull", image: nil, tag: 0)
        var arr = Array<UIViewController>()
        arr.append(pullController)
        setViewControllers(arr, animated: true)
    }
    
    override func viewDidAppear(_ animated: Bool) {
        super.viewDidAppear(animated)
        tabBarOriginFrame = tabBar.frame
        willTabBarHidden = tabBar.isHidden
    }
    
    func setTabBar(hidden:Bool,animation:Bool) {
        if tabBar.isHidden == hidden {
            return
        }
        if willTabBarHidden == hidden {
            return
        }
        willTabBarHidden = hidden
        func handle() {
            if hidden {
                let frame : CGRect = tabBar.frame
                tabBar.frame = CGRect(x: frame.origin.x, y: tabBarOriginFrame.origin.y+frame.size.height, width: frame.size.width, height: frame.size.height)
                tabBar.alpha = 0
            }else{
                let frame : CGRect = tabBar.frame
                tabBar.frame = CGRect(x: frame.origin.x, y: tabBarOriginFrame.origin.y, width: frame.size.width, height: frame.size.height)
                tabBar.alpha = 1
            }
        }
        if animation {
            UIView.animate(withDuration: 0.1, animations: {
                handle()
            }) { (finish) in
                self.tabBar.isHidden = hidden
            }
        }
    }
}

extension TabBarController:PullContainerViewControllerDelegate {
    func pullContainerViewController(controller: PullContainerViewController, progress: CGFloat) {
        if controller.status == .top {
            if !tabBar.isHidden {
                setTabBar(hidden: true, animation: true)
            }
        }else{
            if tabBar.isHidden {
                setTabBar(hidden: false, animation: true)
            }
        }
    }
}
